build_tools/update_iree_requirement_pins.py

```diff
@@ -0,0 +1,195 @@
+    print(f"Finding latest available package version for package '{package_name}'\n")
+
+    # This queries the pip index to get the latest version.
+    #
+    # This doesn't require downloading or installing, so it should be more
+    # portable across operating systems than other approaches.
+    #
+    # Note: the `index` subcommand is experimental. We could instead:
+    #   * Install (into a venv) then check what was installed with `--report`,
+    #     `pip freeze`, or regex parsing (as in `get_current_version()` above)
+    #   * Download then check what was downloaded
+    #   * Scrape the package index and/or release page (https://iree.dev/pip-release-links.html)
+    subprocess_args = [
+        sys.executable,
+        "-m",
+        "pip",
+        "index",
+        "versions",
+        package_name,
+        "--disable-pip-version-check",
+    ]
+    subprocess_args.extend(extra_pip_args)
+
+    print(f"Running command:\n  {subprocess.list2cmdline(subprocess_args)}\n")
+    result = subprocess.run(subprocess_args, stdout=subprocess.PIPE)
+    output = result.stdout.decode("utf-8")
+    print(f"Command output:\n{textwrap.indent(output, '  ')}")
+
+    # Search for text like `iree-base-compiler (3.2.0rc20250109)` within the
+    # multiple lines of output from the command.
+    # WARNING: The output from `pip index` is UNSTABLE and UNSTRUCTURED, but
+    # this seems to work using Python 3.11.2 and pip 22.3.1.
+    version_search_regex = re.compile(f"{package_name}\s\((.*)\)")
+    matches = version_search_regex.match(output)
+    if not matches:
+        raise RuntimeError("Failed to find a package version using regex")
+    version = matches.groups()[0]
+    print(
+        f"Found package version for '{package_name}' in output using regex: '{version}'"
+    )
+    return version
+
+
+def get_current_git_tag():
+    with open(SHORTFIN_CMAKELISTS_PATH, "r") as f:
+        text = f.read()
+        return re.findall('SHORTFIN_IREE_GIT_TAG "(.*)"', text)[0]
+
+
+def main():
+    print("Updating IREE version pins!")
+
+    current_compiler_version = get_current_version("iree-base-compiler")
+    current_runtime_version = get_current_version("iree-base-runtime")
+    current_turbine_version = get_current_version("iree-turbine")
+    current_git_tag = get_current_git_tag()
+
+    nightly_pip_args = [
+        "--pre",
+        "--find-links",
+        "https://iree.dev/pip-release-links.html",
+    ]
+    latest_compiler_version = get_latest_version("iree-base-compiler", nightly_pip_args)
+    latest_runtime_version = get_latest_version("iree-base-runtime", nightly_pip_args)
+    latest_turbine_version = get_latest_version("iree-turbine", nightly_pip_args)
+    # TODO(scotttodd): Get this from git? It should generally be in sync with
+    #     the python packages and follow a naming convention. If that isn't
+    #     true, such as right after a stable release, then this may break.
+    latest_git_tag = f"iree-{latest_runtime_version}"
```

Comment:
I thought through this a little longer and I think there is no reason this should not be in sync, even if patch releases for iree-base-compiler get published. The tag + the latest wheel version should always be in sync.

Reply:
Consider the 3.2.0 release. There will be a window between when we publish 3.2.0 to PyPI and when we publish the first 3.3.0rcYYYYMMDD release during which the latest version fetched by this script will be 3.2.0. This will try to use the git tag `iree-3.2.0` which will not exist if we stick to the new `vX.Y.Z` tag naming (e.g. https://github.com/iree-org/iree/releases/tag/v3.1.0). That would not have been an issue back when we used `iree-X.Y.Z` (e.g. https://github.com/iree-org/iree/releases/tag/iree-3.0.0)

Reply:
One workaround would be to change the `get_latest_version()` function to scrape https://iree.dev/pip-release-links.html instead of calling `pip index`. The release links page _only_ contains prereleases, while `pip index` searches that page _and_ PyPI.

I'm not too concerned about this edge case though, since we'll be paying pretty close attention to the repositories and package versions around when we push a stable release. Where it will matter is if we decide to change the tag naming convention in the future, since the script expects the `iree-PACKAGE_VERSION` naming scheme.

Reply:
We can also push `iree-X.Y.Z` + `vX.Y.Z` for stable releases ;)